### PR TITLE
Update triton pin

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126098

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @desertfire @chauhang

### CODE REVIEW

**Comment by eellison** [comment]
cc @bertmaher for triton pin update

**Comment by eellison** [comment]
cc @desertfire triton oncall

**Comment by bertmaher** [comment]
We've been meaning to cut a new Triton release for PT 2.4 soon anyways, and this commit is probably as good as any.  Let me kick off a more extensive round of tests on fbinfra.

In the meantime, what's up with the test failures?  I can't tell if they're flaky (some of the hf ones looks suspiciously flaky), but some of them might be legit?

**Comment by amjames** [comment]
@bertmaher I was working on resolving a BC breaking change in triton and fixing our code-gen for that situation (see the rest of the stack), but I should now have time to dig into the other failures. 

- I think the wheel build failures are because there is no published triton wheel for the new version. I am not sure If that is something I can fix or if it will be resolved after the pin update lands and nightlies for the new triton version are pushed.

- The HF failure is suspicious: in some cases it is failing due to more graph breaks than expected. In another case it is passing the accuracy check when it is expected that it would fail to run. I will take a closer look at that to make sure. 

Need to investigate the the other two.

**Comment by peterbell10** [comment]
> I think the wheel build failures are because there is no published triton wheel for the new version. I am not sure If that is something I can fix or if it will be resolved after the pin update lands and nightlies for the new triton version are pushed.

If you mean this error:
```
ERROR: Could not find a version that satisfies the requirement pytorch-triton==3.0.0+afaf1f0491
```
then yes this happens every pin update because CI on PRs doesn't upload the wheel. After it gets merged to main this should succeed.

> The HF failure is suspicious: in some cases it is failing due to more graph breaks than expected. In another case it is passing the accuracy check when it is expected that it would fail to run. I will take a closer look at that to make sure.

`hf_BigBird` has been failing in CI on all PRs recently, in case that's what you're seeing.

**Comment by shunting314** [comment]
There is another attempt to move the pin: https://github.com/pytorch/pytorch/pull/127429

**Comment by bertmaher** [comment]
Let's try to cut a Triton release branch from 513f38c4 (recommended by @ptillet) and then add the cherry picks on top to fix the demat pass

**Comment by etaf** [comment]
Hi, @amjames may I ask if this triton update will be included in release 2.4?

**Comment by bertmaher** [comment]
Trying to get this in.  I've been working through some meta-internal test failures (so far just errors due to API churn, no Triton bugs... yet), and I think @amjames is working through an error in rematerialization.  @atalman do you think we'll be able to get this version bump into the release?

**Comment by amjames** [comment]
> Hi, @amjames may I ask if this triton update will be included in release 2.4?

@etaf That is the plan for right now. 

I have found and fixed the error in remat, need to write a minimal test which is proving challenging, but I should have that done this week. Once it is landed we will need to move the pin again and see where we are

**Comment by alexbaden** [comment]
Hi @amjames - do you have the remat error or fix description available for me to look at? I think once again we might be independently looking into the same problem. 😀

**Comment by jataylo** [comment]
This change https://github.com/triton-lang/triton/pull/4049 may also want to be considered for cherry-pick, without this I'm seeing some issues on python3.8 on aot inductor UTs.

**Comment by bertmaher** [comment]
@amjames Would you mind changing this PR to bump to 513f38c4?  I've been testing that pin internally with Meta CI and that's a pretty lengthy process so I'd rather use that as a starting point.  We can then create a release branch (@atalman, could you sync with @ptillet for that?) and add the necessary cherry-picks to fix remat and 3.8 support

**Comment by amjames** [comment]
@bertmaher I have bumped the pin to 513f38c4 for now and triton-lang/triton#4097 is open to address the remat issue. 

Is your plan to land this bump without that upstream fix? If that is the case we will have to add a skip for the problematic flex-attention test to get past CI.

**Comment by amjames** [comment]
The upstream fix landed quicker than I thought it would so I bumped the pin agin. Hopefully nothing else comes up. 🤞

**Comment by amjames** [comment]
FYI: triton-lang/triton#4130 will give a route to address the issue with signal propagation. My fix on this branch does not resolve the issue that the stacktrace handler installed by triton is preventing the signal from propagating.

File: torch/_inductor/kernel/mm.py [hunk 1]
```diff
@@ -89,7 +89,7 @@
             b = tl.load(B, mask=rk[:, None] < k, other=0.)
         if B_PROLOGUE_CAST_TYPE is not None:
             b = b.to(B_PROLOGUE_CAST_TYPE)
-        acc += tl.dot(a, b, allow_tf32=ALLOW_TF32)
+        acc += tl.dot(a, b, input_precision=INPUT_PRECISION)
```
Comment (lezcano) [inline torch/_inductor/kernel/mm.py:92]: Moving the conversation here:
As reported by CUTLASS, this trick just gives higher precision for large enough matrices (k >= 200 or so). See https://www.nvidia.com/en-us/on-demand/session/gtcspring22-s41996/?playlistId=playList-d59c3dc3-9e5a-404d-8725-4b567f4dfe77 slide 39
image

I think we should perform a comprehensive benchmark of accuracy and also performance for smaller sizes if we are going to activate this by default.

**Review by bertmaher — APPROVED** [review]

**Pull request closed without merging**